#30255: Fixed "The coupon code has been accepted." message remains after a Coupon Code was removed

shikhamis11 opened the pull request:
Fixed #30255 "The coupon code has been accepted." message remains after a Coupon Code was removed

**Preconditions**

    Create simple product

    Create Customer with address

    First Name: John
    Last Name: Doe
    Email Address: john.doe@example.com(use real mail)
    Company: Magento
    Address: 6161 West Centinela Avenue
    City: Culver City
    State/Province: California (CA)
    Zip/Postal Code: 90230
    Country: United States
    Telephone: 555-55-555-55

    Go to Admin -> Marketing -> Cart Price Rules: click "Add New Rule"

    Fill all required fields
        "Coupon" select Specific Coupon
        "Coupon Code" = 1111
    In "Action" section
        "Apply" select "Percent of product price discount"
        "Discount Amount" = 50
    click "Save" button

**Steps:**

    Go to admin -> sales -order -> create new order -> select customer
    Add Products => simple product
    Apply Coupon Code => 1111
    remove Coupon Code

**(/) Expected result**
"The coupon code has been accepted." message should disappear

**(x) Actual result**
"The coupon code has been accepted." message is still present

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

Commit:
#30255: Fixed "The coupon code has been accepted." message remains after a Coupon Code was removed

Fixed "The coupon code has been accepted." message remains after a Coupon Code was removed

rodrigowebjump commented:
@magento give me test instance

milindsingh commented:
#dmcdindia2020

pramod-mishra commented:
#dmcdindia2020

magento-engcom-team commented:
@pramod-mishra thank you for joining. Please accept team invitation [here](https://github.com/orgs/magento/invitation) and self-assign the issue.

rodrigowebjump commented:
Hi @shikhamis11 Thanks for your contribution.

Please could you cover your fix by autotests (MFTF)? 
Thanks

pramod-mishra commented:
@magento give me test instance

shikhamis11 commented:
@rodrigowebjump Thanks for the review . I am working on to add MFTF coverage for the issue

rodrigowebjump commented:
@magento give me 2.4-develop instance

rodrigowebjump commented:
Hi @shikhamis11 

Apparently the issue is being caused because there is no message returned when the coupon is removed. 
A suggestion for this case is to return a message informing that the coupon was removed in the backend request process (like occurs in the storefront) 

The current JS behavior does not remove the previous message if a empty is received in the response. Since there is a lot of requests happening there, my recommendation is to not change this behavior for now because we can create some undesirable 
result.

shikhamis11 committed:
#30255: Fixed The coupon code has been accepted. message remains after a Coupon Code was removed

reverted previous changes and updated requested changes

shikhamis11 commented:
@magento give me test instance

shikhamis11 commented:
@magento give me test instance

shikhamis11 commented:
@magento run all tests

shikhamis11 committed:
#30255:#30255: Fixed The coupon code has been accepted message

Fixed Static test
Added MFTF Test to Verify Remove coupon code

shikhamis11 commented:
@magento run all tests

gabrieldagama commented on app/code/Magento/Sales/Controller/Adminhtml/Order/Create.php:
```diff
@@ -344,6 +344,8 @@ protected function _processActionData($action = null)
                     $this->messageManager->addSuccessMessage(__('The coupon code has been accepted.'));
                 }
             }
+        } elseif (isset($data['coupon']['code']) && empty($couponCode)) {
+                    $this->messageManager->addSuccessMessage(__('The coupon code has been removed.'));
```
```suggestion
        } elseif (isset($data['coupon']['code']) && empty($couponCode)) {
            $this->messageManager->addSuccessMessage(__('The coupon code has been removed.'));
```

gabrieldagama replied:
```suggestion
        } elseif (isset($data['coupon']['code']) && empty($couponCode)) {
            $this->messageManager->addSuccessMessage(__('The coupon code has been removed.'));
```

gabrieldagama reviewed (commented):
Hi @shikhamis11, thanks for the PR, we just need to adjust the code formatting.

gabrieldagama reviewed (changes requested):
Changing to the right column, waiting for contributor changes.

Commit:
Updated suggestes changes

Co-authored-by: Gabriel da Gama <gabriel@gabrielgama.com.br>

shikhamis11 commented:
@magento run all tests

gabrieldagama commented on app/code/Magento/Sales/Test/Mftf/Test/AdminCreateOrderToVerifyApplyAndRemoveCouponCodeTest.xml:
```diff
@@ -0,0 +1,100 @@
+            <magentoCLI command="config:set {{EnablePaymentBankTransferConfigData.path}} {{EnablePaymentBankTransferConfigData.value}}" stepKey="enableBankTransferPayment"/>
+
+            <!--Set default flat rate shipping method settings-->
+            <magentoCLI command="config:set {{EnableFlatRateConfigData.path}} {{EnableFlatRateConfigData.value}}" stepKey="enableFlatRate"/>
```
Don't we need to revert this state as well on after node?

rodrigowebjump replied:
I think that in this case you can use the following 

<createData entity="FlatRateShippingMethodDefault" stepKey="setDefaultFlatRateShippingMethod"/>

as used in 
https://github.com/magento/magento2/blob/1ee91ca2d9c6650c83f4b5f6abc1983fa1fef629/app/code/Magento/Sales/Test/Mftf/Test/AdminCreateOrderWithSimpleProductTest.xml

https://github.com/magento/magento2/blob/1ee91ca2d9c6650c83f4b5f6abc1983fa1fef629/app/code/Magento/Sales/Test/Mftf/Test/AdminHoldCreatedOrderTest.xml


@gabrieldagama I don't see the revert of this flatrate set in other tests. Is this a good practice in this case or is not required?

rodrigowebjump replied:
@shikhamis11 You can use the files I mentioned as a reference for before actions.

gabrieldagama replied:
@rodrigowebjump Yeah, ideally we should leave the application state as we found before the test ran, the reason for that is basically to avoid flaky tests. Other tests may rely on the "default" application state to pass successfully, and that may bring problems for us on PR processing.

gabrieldagama commented on app/code/Magento/Sales/Test/Mftf/Test/AdminCreateOrderToVerifyApplyAndRemoveCouponCodeTest.xml:
```diff
@@ -0,0 +1,100 @@
+        </actionGroup>
+        
+        <!-- Remove coupon from the order -->
+        <actionGroup ref="AdminRemoveCouponFromOrderActionGroup" stepKey="removeCoupon"/>
```
I guess we need to assert that the message old message disappear and the new one shows up

shikhamis11 committed:
#30255: Fixed The coupon code has been accepted message

Added after action for disable Flatrate in MFTF test
Added Verification for previous message removed

shikhamis11 commented:
@engcom-Charlie yes it is ready 
I am waiting for review from @rodrigowebjump and @gabrieldagama

engcom-Charlie commented:
@magento run all tests

gabrieldagama commented on app/code/Magento/Sales/Test/Mftf/Test/AdminCreateOrderToVerifyApplyAndRemoveCouponCodeTest.xml:
```diff
@@ -0,0 +1,112 @@
+            <group value="sales"/>
+        </annotations>
+        <before>
+            <!-- Login as admin -->
```
Please remove the comments between the steps, they are pretty much the same content already on the stepKey attribute, I don't think it brings any benefit.

gabrieldagama reviewed (changes requested):
Hi @shikhamis11, thanks for updating the PR, can you have a look at my comment below? It is a pretty small thing but would be nice to keep the test cleaner.

shikhamis11 committed:
#30255: Fixed The coupon code has been accepted message

Removed comments

gabrieldagama commented:
@magento run all tests

magento-engcom-team commented:
Hi @gabrieldagama, thank you for the review.
[ENGCOM-8426](https://jira.corp.magento.com/browse/ENGCOM-8426) has been created to process this Pull Request
:eight_spoked_asterisk: @gabrieldagama, could you please add one of the following labels to the Pull Request?

| Label | Description |
|--------------------------------|-----------------------------------------------------------------|
| ```Auto-Tests: Covered``` | All changes in Pull Request is covered by auto-tests |
| ```Auto-Tests: Not Covered``` | Changes in Pull Request requires coverage by auto-tests |
| ```Auto-Tests: Not Required``` | Changes in Pull Request does not require coverage by auto-tests |

engcom-Alfa commented:
:heavy_check_mark: QA Passed

**Preconditions:**

1. Create Simple product;
2. Create Customer;
3. Go to Admin -> Marketing -> Cart Price Rules: click "Add New Rule"
```
Fill all required fields
    "Coupon" select Specific Coupon
    "Coupon Code" = 12345
In "Action" section
    "Apply" select "Percent of product price discount"
    "Discount Amount" = 50
click "Save" button
```

**Manual testing scenario:**

1. Go to admin -> sales -> orders -> create new order -> select customer
2. Add Products => simple product
3. Apply Coupon Code => 12345
4. remove Coupon Code

**Before:** :heavy_multiplication_x:  "The coupon code has been accepted." message is still present

![Peek 2020-11-05 09-49](https://user-images.githubusercontent.com/51679138/98214457-3232ce80-1f4f-11eb-8128-16e97e0be742.gif)

**After:** :heavy_check_mark: "The coupon code has been removed." message is present

![Peek 2020-11-05 10-00](https://user-images.githubusercontent.com/51679138/98214591-5d1d2280-1f4f-11eb-8ea8-4ed5dbd91a36.gif)

engcom-Charlie commented on app/code/Magento/Sales/Test/Mftf/Test/AdminCreateOrderToVerifyApplyAndRemoveCouponCodeTest.xml:
```diff
@@ -0,0 +1,72 @@
+            <stories value="Create Order with offline payment methods"/>
+            <title value="Create Order to verify apply and remove coupon code test"/>
+            <description value="Create Order to verify apply and remove coupon code test"/>
+            <severity value="CRITICAL"/>
+            <testCaseId value="MC-37386"/>
```
@shikhamis11 could you please update `severity` and `testCaseId`?
Thank you.
```
            <severity value="AVERAGE"/>
            <testCaseId value="MC-38919"/>
```

Commit:
#30255: Fixed The coupon code has been accepted message

updated `severity` and `testCaseId` for added MFTF test

engcom-Charlie commented:
@magento run all tests

gabrieldagama commented on app/code/Magento/Sales/Test/Mftf/Test/AdminCreateOrderToVerifyApplyAndRemoveCouponCodeTest.xml:
```diff
@@ -0,0 +1,72 @@
+            <magentoCLI
+                command="config:set {{DisablePaymentBankTransferConfigData.path}} {{DisablePaymentBankTransferConfigData.value}}"
+                stepKey="disableBankTransferPayment"/>
+            <magentoCLI command="config:set {{DisableFlatRateConfigData.path}} {{DisableFlatRateConfigData.value}}"
+                        stepKey="disableFlatRate"/>
```
I believe flat rate is enabled by default, by disabling it we may have made other tests flaky. I guess you can keep the CLI to enable it, but remove this one.

gabrieldagama reviewed (changes requested):
Hi @shikhamis11, I believe we might have an issue with the mftf, can you check it?

Commit:
#30255: Fixed The coupon code has been accepted message

removed disable flat rate step from after action

shikhamis11 commented:
@magento run all tests

gabrieldagama commented:
@magento run Functional Tests EE

magento-engcom-team commented:
Hi @gabrieldagama, thank you for the review.
[ENGCOM-8426](https://jira.corp.magento.com/browse/ENGCOM-8426) has been created to process this Pull Request

engcom-Alfa commented:
:heavy_check_mark: QA Passed
